File: src/main/java/mc/fbturrets/main/FBTurrets.java
package mc.fbturrets.main;

import mc.fbturrets.blocks.TurretGun;
import mc.fbturrets.blocks.TurretHolder;
import mc.fbturrets.blocks.TurretHolderBlockEntity;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.object.builder.v1.block.FabricBlockSettings;
import net.minecraft.block.Block;
import net.minecraft.block.Material;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.particle.DefaultParticleType;
import net.minecraft.particle.ParticleTypes;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.Box;
import net.minecraft.util.registry.Registry;

public class FBTurrets implements ModInitializer {
	public static final String MOD_ID = "fbturrets";

	public static final Identifier TURRET_ANGLES_UPDATE = new Identifier(MOD_ID, "turret_angles");
	public static final Identifier TURRET_SHOOT_PARTICLES = new Identifier(MOD_ID, "turret_shoot");

	public static Block TURRET_HOLDER_MODEL;
	public static TurretGun BASIC_GUN;
	public static TurretGun SNIPER_GUN;

	public static final TurretHolder TURRET_HOLDER = new TurretHolder(FabricBlockSettings.of(Material.BARRIER).hardness(4.0f).nonOpaque());
	public static BlockEntityType<TurretHolderBlockEntity> TURRET_HOLDER_BLOCK_ENTITY;

	@Override
	public void onInitialize() {
		TURRET_HOLDER_MODEL = new Block(FabricBlockSettings.of(Material.METAL).hardness(4.0f).nonOpaque());
		Registry.register(Registry.BLOCK, new Identifier(MOD_ID, "turret_holder_render_only"), TURRET_HOLDER_MODEL);

		BASIC_GUN = registerTurretGun("basic_gun", 0.3, 20, ParticleTypes.CRIT, 5, new Box(10, 5, 10, -10, -5, -10));
		SNIPER_GUN = registerTurretGun("sniper_gun", 0.2, 35, ParticleTypes.ENCHANTED_HIT, 15, new Box(32, 15, 32, -32, -15, -32));

		Registry.register(Registry.BLOCK, new Identifier(MOD_ID, "turret_holder"), TURRET_HOLDER);
		TURRET_HOLDER_BLOCK_ENTITY = Registry.register(Registry.BLOCK_ENTITY_TYPE, MOD_ID + ":turret_holder", BlockEntityType.Builder.create(TurretHolderBlockEntity::new, TURRET_HOLDER).build(null));
		Registry.register(Registry.ITEM, new Identifier(MOD_ID, "turret_holder"), new BlockItem(TURRET_HOLDER, new Item.Settings().group(ItemGroup.MISC)));
	}

	private TurretGun registerTurretGun(String ident, double aimSpeed, int aimTime, DefaultParticleType effect, int damage, Box box) {
		TurretGun block = new TurretGun(aimTime, aimSpeed, damage, effect, box, FabricBlockSettings.of(Material.METAL).hardness(4.0f).nonOpaque());
		Registry.register(Registry.BLOCK, new Identifier(MOD_ID, ident), block);
		return block;
	}
}

File: src/main/kotlin/mc/fbturrets/main/FBTurrets.kt
package mc.fbturrets.main

import mc.fbturrets.blocks.TurretGun
import mc.fbturrets.blocks.TurretHolder
import mc.fbturrets.blocks.TurretHolderBlockEntity
import mc.fbturrets.blocks.gui.TurretHolderGuiDescription
import mc.fbturrets.items.GunItem
import mc.fbturrets.util.MathHelp
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.`object`.builder.v1.block.FabricBlockSettings
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder
import net.fabricmc.fabric.api.screenhandler.v1.ScreenHandlerRegistry
import net.minecraft.block.Block
import net.minecraft.block.Material
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.item.*
import net.minecraft.particle.DefaultParticleType
import net.minecraft.particle.ParticleTypes
import net.minecraft.screen.ScreenHandlerContext
import net.minecraft.screen.ScreenHandlerType
import net.minecraft.util.Identifier
import net.minecraft.util.math.Box
import net.minecraft.util.registry.Registry
import java.util.function.Supplier

class FBTurrets : ModInitializer {
    override fun onInitialize() {
        TURRET_HOLDER_MODEL = Block(FabricBlockSettings.of(Material.METAL).hardness(4.0f).nonOpaque())
        Registry.register(Registry.BLOCK, Identifier(MOD_ID, "turret_holder_render_only"), TURRET_HOLDER_MODEL)
        Registry.register(Registry.BLOCK, Identifier(MOD_ID, "turret_holder"), TURRET_HOLDER)
        TURRET_HOLDER_BLOCK_ENTITY = Registry.register(Registry.BLOCK_ENTITY_TYPE, "$MOD_ID:turret_holder", BlockEntityType.Builder.create(Supplier { TurretHolderBlockEntity() }, TURRET_HOLDER).build(null))
        Registry.register(Registry.ITEM, Identifier(MOD_ID, "turret_holder"), BlockItem(TURRET_HOLDER, Item.Settings().group(ITEM_GROUP)))
        BASIC_GUN = registerTurretGun("basic_gun", 0.3, 20, ParticleTypes.CRIT, 5, MathHelp.buildSimpleBox(15))
        SNIPER_GUN = registerTurretGun("sniper_gun", 0.2, 35, ParticleTypes.ENCHANTED_HIT, 15, MathHelp.buildSimpleBox(32))
        AUTO_GUN = registerTurretGun("auto_gun", 0.35, 11, ParticleTypes.CLOUD, 3, MathHelp.buildSimpleBox(10))
        TURRET_HOLDER_SCREEN_TYPE = ScreenHandlerRegistry.registerSimple(Identifier(MOD_ID, "turret_holder")) { syncId, inventory-> TurretHolderGuiDescription(syncId, inventory, ScreenHandlerContext.EMPTY) }
    }

    private fun registerTurretGun(ident: String, aimSpeed: Double, aimTime: Int, effect: DefaultParticleType, damage: Int, box: Box): TurretGun {
        val id = Identifier(MOD_ID, ident)
        val block = TurretGun(aimTime, aimSpeed, damage, effect, box, id, FabricBlockSettings.of(Material.METAL).hardness(4.0f).nonOpaque())
        Registry.register(Registry.BLOCK, id, block)
        Registry.register(Registry.ITEM, id, GunItem(id, Item.Settings().group(ITEM_GROUP)))
        return block
    }

    companion object {
        const val MOD_ID = "fbturrets"
        val ITEM_GROUP: ItemGroup = FabricItemGroupBuilder.build(
                Identifier(MOD_ID, "item_group")
        ) { ItemStack(Items.CROSSBOW) }
		val TURRET_ANGLES_UPDATE = Identifier(MOD_ID, "turret_angles")
		val TURRET_SHOOT_PARTICLES = Identifier(MOD_ID, "turret_shoot")
		val TURRET_GUN_UPDATE = Identifier(MOD_ID, "turret_update")
		lateinit var TURRET_HOLDER_MODEL: Block
        lateinit var BASIC_GUN: TurretGun
        lateinit var SNIPER_GUN: TurretGun
        lateinit var AUTO_GUN: TurretGun
        @JvmField
		val TURRET_HOLDER = TurretHolder(FabricBlockSettings.of(Material.BARRIER).hardness(4.0f).nonOpaque())
		lateinit var TURRET_HOLDER_BLOCK_ENTITY: BlockEntityType<TurretHolderBlockEntity>
        lateinit var TURRET_HOLDER_SCREEN_TYPE : ScreenHandlerType<TurretHolderGuiDescription>
    }
}